cpu/atmega256rfr2/periph/cpuid.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2018 RWTH Aachen, Josua Arndt, Steffen Robertz
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @addtogroup  cpu_atmega256rfr2
+ * @{
+ *
+ * @file
+ * @brief       Low-level CPUID driver implementation
+ *
+ * @author      Steffen Robertz <steffen.robertz@rwth-aachen.de>
+ * @author      Josua Arndt <jarndt@ias.rwth-aachen.de>
+ *
+ * @}
+ */
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "periph/cpuid.h"
+
+#include "atmega_regs_common.h"
+#include "avr/boot.h"
+#ifdef MODULE_AT86RF2XX
+    #include "at86rf2xx_registers.h"
+#endif
+
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+/**
+ * @brief CPU_ID build from MCU register
+ *
+ *  CPIUD is taken from MCU Control Register and Signature bytes.
+ *  CPUID:  1e  a8  02  ff  ff  1f  94  03 XX XX
+ *  MEGA62/128/256_RFR2 [MANUAL] p.138
+ *  MEGA62/128/256_RFR2 [MANUAL] p.507
+ *
+ *  usr_sign_0/1 are configurable values on flash page 1.
+ *
+ *  A random 16bit ending can be added when AT86RFR2 Module is present.
+ *  This can be configured in board/include/periph_cpu.h
+ */
+void cpuid_get(void *id)
+{
+
+    uint8_t  signature_0 = boot_signature_byte_get(0x00);
+    uint8_t  signature_1 = boot_signature_byte_get(0x02);
+    uint8_t  signature_2 = boot_signature_byte_get(0x04);
+
+    uint8_t  usr_sign_0 = boot_signature_byte_get(0x0100);
+    uint8_t  usr_sign_1 = boot_signature_byte_get(0x0102);
+
+    /* use random number generator to get last byte for the ID */
+    uint16_t random;
+
+#if CPU_ID_ADD_RANDOME
+    /* support will be added once the driver is pushed, until then commented out */
+    /* TODO: use random generated by hardware */
+    at86rf2xx_get_random_num( &random, 2);
+#else
```

Comment:
The CPUID should be static (and unique) on the device basis. I would like to handle the random component outside this feature

Reply:
Does this MCU provide unique IDs?

Reply:
Problem is that the atmega256rfr2 does not have a unique ID. Thats why we tried to build own ones. But as all the information we have (like part number, revision, etc) are the same between the jiminy boards, we added the random function. Another option would be to generate this random ID once and save it in the EEPROM. Then we could just read it back using the cpuid function. At least it would return the same ID on consecutive calls of the ID function

Reply:
AS the CPUID is used to generate the IPV6 we added a randome to get different IPV6 on different boards.
The CPUID is constructed out of MCU specific values, plus the used transciever if there is one.
So it is basically like RPIs the prefix is fix and the ending is different.

Reply:
Understand. But in the current state, this won't work anyway. So can we remove it for now? "Some time in future" we'd like to provide a more generic way for such ID generation. Right now we just configure a fallback ID which can be overwritten at compile time for MCUs without unique vendor IDs. Furthermore one can still set "random-based" addresses during runtime. Would this suffice your needs in a first approach?

Reply:
The CPIUD is now static, only the last two bytes which can be set on flash page 1 to enable additional information placement.

Reply:
I din not find Information about how to set up cpuid, the other boards like sam0_common just define values without any additional information from where they are. The other boards have CPUID implemented in the MCU, or just use TPU_DEVID. 
So I think using The Atmel manufacturer ID, Device Identification Part Number,  Atmel JTAG manufacturer ID, PART NUM, VERSION NUM, , If applied also add informaiton about the transceiver , and some user specific information, are a good CPUID.

Is there a RIOT CPUID definition? From where the build up of the CPIUD can be taken?